### PR TITLE
Equivariant models

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
~~This allows for sharing parameters inter and intra circuit layers. Namely, the independent parameters are extracted from the circuit using the [np.shares_memory](https://numpy.org/doc/stable/reference/generated/numpy.shares_memory.html) method and added as parameters of the model. Upon circuit reconstruction, then, a new custom ``set_parameters`` function is used to duplicate the independent parameters and match the circuit length.~~

EDIT: In the end I decided to go for a simpler and cleaner solution, inspired by ``pennylane``'s approach to this matter.
Namely, I'm allowing the ``QuantumModel`` object to accept even `Callable` functions that construct a circuit given some input parameters (it is in fact very similar to an encoder, with the sole difference that the inputs are parameters of the model themselves). Therefore, to construct equivariant circuits, the user needs to define a function of the independent parameters that builds the circuit in object, e.g.:
```python
    # this defines 3 independent parameters
    def custom_circuit(engine, th, phi, lam):
        c = Circuit(2)
        delta = 2 * engine.cos(phi) + lam**2
        gamma = lam * engine.exp(th / 2)
        c.add([gates.RZ(i, theta=th) for i in range(2)])
        c.add([gates.RX(i, theta=lam) for i in range(2)])
        c.add([gates.RY(i, theta=phi) for i in range(2)])
        c.add(gates.RZ(0, theta=delta))
        c.add(gates.RX(1, theta=gamma))
        return c
```
Note that an additional `engine` argument is needed which will cover the role of `numpy`/`torch`/`tensorflow` or any engine to perform operations on the parameters. This is needed because at the moment we defer the circuit construction to the moment we calculate the custom gradients (with our `qiboml.differentiation.jax` and `qiboml.differentiation.PSR`) which then requires an engine compatible with the backend to be used  inside the custom function. 
This need could be lifted by directly constructing the whole circuit in the `interface`'s API (thus `torch` or `keras`), and delegating the calculation of the derivatives of any possible operations performed on the independent parameters to the interface as well, whereas the actual derivative wrt the circuit angles will be taken care of by our custom differentiation engine and correctly combined following the chain rule.  This is probably the best solution, as in order to make PSR work with equivariant models we would need to set up a similar structure anyway (thus at the moment PSR with equivariant models is not supported in this PR). However, this approach may involve a substantial refactor of our whole custom differentiation pipeline, which is maybe better to keep separated from this.
As a final note, keep in mind that each independent parameter has to be defined separately in the function and it's always inferred to be a scalar, since I use `inspect.signature` to retrieve the number of independent parameters and there would be no way (I believe) to infer the correct shape of each parameter from that.

That said, then the `QuantumModel` can be constructed as usual:
```python
model = QuantumModel([circuit, encoding, custom_circuit], decoding)
```

This should help with #108 and #99.

EDIT 2: with the new refactor #140 merged here the need for the engine to be passed to the function has been lifted, the appropriate engine for the interface should be used now. Thus, with ``torch``:
```python
    # this defines 3 independent parameters
    def custom_circuit(th, phi, lam):
        c = Circuit(2)
        delta = 2 * torch.cos(phi) + lam**2
        gamma = lam * torch.exp(th / 2)
        c.add([gates.RZ(i, theta=th) for i in range(2)])
        c.add([gates.RX(i, theta=lam) for i in range(2)])
        c.add([gates.RY(i, theta=phi) for i in range(2)])
        c.add(gates.RZ(0, theta=delta))
        c.add(gates.RX(1, theta=gamma))
        return c
```
whereas with ``keras``:
```python
    # this defines 3 independent parameters
    def custom_circuit(th, phi, lam):
        c = Circuit(2)
        delta = 2 * keras.ops.cos(phi) + lam**2
        gamma = lam * keras.ops.exp(th / 2)
        c.add([gates.RZ(i, theta=th) for i in range(2)])
        c.add([gates.RX(i, theta=lam) for i in range(2)])
        c.add([gates.RY(i, theta=phi) for i in range(2)])
        c.add(gates.RZ(0, theta=delta))
        c.add(gates.RX(1, theta=gamma))
        return c
```

### CODE REVIEW

**Review by MatteoRobbiati — COMMENTED** [review]
Thanks a lot for this! A few comments follow, but we can go ahead.
